pkg/tenant/db/members.go

```diff
@@ -52,11 +53,73 @@ func (m *Member) UnmarshalValue(data []byte) error {
 	return msgpack.Unmarshal(data, m)
 }
 
-// CreateMember adds a new Member to the database.
+func (m *Member) Validate() error {
+	memberName := m.Name
+
+	if strings.ContainsAny(string(memberName[0]), "0123456789") {
+		return ErrNumberFirstCharacter
+	}
+
+	if strings.ContainsAny(memberName, " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~") {
+		return ErrSpecialCharacters
+	}
+
+	return nil
+}
+
+// Validates data in the member model when a TenantID is required.
+// Ex. CreateTenantProject and UpdateProject
+func (m *Member) ValidateWithID() error {
+	if m.TenantID.Compare(ulid.ULID{}) == 0 {
+		return ErrMissingTenantID
+	}
+
+	memberName := m.Name
+
+	if strings.ContainsAny(string(memberName[0]), "0123456789") {
+		return ErrNumberFirstCharacter
+	}
+
+	if strings.ContainsAny(memberName, " !\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~") {
+		return ErrSpecialCharacters
+	}
+
+	return nil
+}
+
+// CreateTenantMember adds a new Member to a tenant in the database.
+// Note: If a memberID is not passed in by the User, a new member id will be generated.
+func CreateTenantMember(ctx context.Context, member *Member) (err error) {
+	// TODO: Use crypto rand and monotonic entropy with ulid.New
+
+	// Validate tenant project data.
+	if err = member.ValidateWithID(); err != nil {
+		return err
+	}
+
+	if member.ID.Compare(ulid.ULID{}) == 0 {
+		member.ID = ulid.Make()
+	}
```

Comment:
Or maybe we could instead move this check above the validation check to ensure that the member always has an ID before it's validated?

Reply:
In the case of `CreateTenantMember` and `CreateTenantProject`, I think we should check if the `TenantID` exists before checking if a `MemberID` or `ProjectID` exists.

Reply:
The `MemberID` check has been moved up before the validation check.